c7n/resources/ec2.py

```diff
@@ -384,6 +385,37 @@ def process_instance(i):
 
         with self.executor_factory(max_workers=2) as w:
             list(w.map(process_instance, instances))
+
+@actions.register('create-snapshot')
+class Snapshots(BaseAction):
+    def process(self, resources):
+        volume_sets = {}
+        for resource in resources:
```

Comment:
instead of building a large in mem structure of all the instances to volume mapping, just pass the instance to process_volume_set and build a transient structure there. that also allows to copy additional instance properties to the snapshots.